packages/host/app/services/loader-service.ts

```diff
@@ -32,8 +41,70 @@ export default class LoaderService extends Service {
       new URL(baseRealm.url),
       new URL(config.resolvedBaseRealmURL),
     );
+    loader.prependURLHandlers([(req) => this.fetchWithAuth(req)]);
     shimExternals(loader);
 
     return loader;
   }
+
+  private async fetchWithAuth(request: Request) {
+    // To avoid deadlock, we can assume that any GET requests to baseRealm don't need authentication.
+    let isGetRequestToBaseRealm =
+      request.url.includes(baseRealm.url) && request.method === 'GET';
+    if (
+      isGetRequestToBaseRealm ||
+      request.url.includes('session') ||
+      request.method === 'HEAD' ||
+      request.headers.has('Authorization')
+    ) {
+      return null;
+    }
+    let realmURL = await this.realmInfoService.fetchRealmURL(request.url);
+    if (!realmURL) {
+      return null;
+    }
+
+    // We have to get public readable status
+    // before we instatiate realm resource and load realm token.
+    // Because we don't want to do authentication
+    // for GET request to publicly readable realm.
+    let isPublicReadable = await this.realmInfoService.isPublicReadable(
+      realmURL,
+    );
+    if (request.method === 'GET' && isPublicReadable) {
+      return null;
+    }
+
+    let realmResource = await this.getRealmResource(realmURL);
+    if (!realmResource.realmToken) {
+      return null;
+    }
+
+    request.headers.set('Authorization', realmResource.realmToken);
+    let body;
+    if (request.bodyUsed) {
+      body = null;
+    } else if (request.headers.get('content-type') === 'application/json') {
+      body = JSON.stringify(await request.json());
+    } else {
+      body = await request.text();
+    }
+    return await this.loader.fetch(request.url, {
+      method: request.method,
+      headers: new Headers(request.headers),
+      body,
+    });
+  }
+
+  private async getRealmResource(realmURL: URL) {
```

Comment:
Would it be better to rename this to `getRealmSession` or `getRealmSessionResource`?